### PR TITLE
refactor(exceptions): raise the periodicity and dimension errors the library defines (#904)

### LINKED ISSUE #904: TIDY_UP-11: Retire or wire the exported-but-unused surface
## Files
- `src/climate_indices/utils.py` — `__all__` (:16-25); zero production callers: `compute_days` (:31), `sign_change` (:130), `is_data_valid` (:159), `rmse` (:186); single caller: `count_zeros_and_non_missings` (:78 → `compute.py:339`), `get_tolerance` (:117 → `__main__.py`), `gregorian_length_as_366day` (:324 → `__spi__.py:587`)
- `src/climate_indices/exceptions.py` — `DimensionMismatchError` (:150), `PeriodicityError` (:242), `BetaFeatureWarning` (:418), `emit_deprecation_warning` (:468); all exported, none raised/called in `src/`
- `tests/test_utils.py` — pins the unused functions

## Problem
The library exports API with no production consumer, and defines exception types no code raises — callers catching `PeriodicityError` or `DimensionMismatchError` silently miss real errors that arrive as another type. `emit_deprecation_warning`/`BetaFeatureWarning` are public and documented but unused, so their contract is exercised only by their own tests.

## Solution
Decide per item and record the decision:
- wire `PeriodicityError` at periodicity validation raises (`compute.py:138,146,174,194,207,271,281`; `indices.py:143`) and `DimensionMismatchError` at shape mismatches, or delete them;
- inline the single-caller utils into their callers, or document them as deliberate public helpers;
- unused utils: deprecate before removal (public API compat — do not delete outright);
- keep or drop `emit_deprecation_warning`/`BetaFeatureWarning` explicitly.

Prefer wiring exceptions first: it improves catch behavior and is directly testable.

## Benefits
Locality: an exception type that exists is one that gets raised. Deletion shrinks the test surface. No more silent catch-misses.

## Recommendation strength
**Worth exploring**

## Related
#662 (bare ValueError + logging migration — overlaps the wiring work), #853 (closed; prior dead-helper removal), #847

## Acceptance
- Any removed public name goes through a documented deprecation step.
- `uv run ruff check src/ tests/`, `uv run ruff format --check src/ tests/`, `uv run mypy src/`, `uv run pytest` pass.

### PR DESCRIPTION
**Pull request by monocongo**
Closes #904

#904's premise re-validated against `origin/main` first: `PeriodicityError` and
`DimensionMismatchError` are exported from `climate_indices.exceptions` and
covered by the contract tables in `tests/test_exceptions.py`, but no module in
`src/` raised either one. Callers catching them missed real failures that
arrived as `ValueError` or `CoordinateValidationError`.

## What this changes

**`PeriodicityError`** now covers the user-supplied periodicity checks:

| Site | Before | After |
| --- | --- | --- |
| `compute._validate_array` (missing, unsupported) | `ValueError` | `PeriodicityError` |
| `compute._reshape_time_major`, `compute.reshape_values` | `ValueError` | `PeriodicityError` |
| `compute.gamma_parameters`, `compute.prepare_scaled` | `ValueError` | `PeriodicityError` |
| `indices._validate_periodicity` | `InvalidArgumentError` | `PeriodicityError` |

Messages, `reason=` log events, and the `argument_name="periodicity"` /
`argument_value` context are unchanged. `PeriodicityError` derives from
`InvalidArgumentError`, so handlers catching that documented parent keep
working; `except ValueError` around these calls no longer catches, which the
migration note records.

**`DimensionMismatchError`** is now a specialization of
`CoordinateValidationError` — the same pattern `PeriodicityError` already has
with `InvalidArgumentError` — and is raised where a *required named dimension*
is missing: `validation.validate_time_dimension` and the HDW `level_dim` check
in `fire/_hdw.py`. Its own docstring already scoped it to "xarray named
dimensions", which is exactly these two sites. `coordinate_name` and
`reason="missing_dimension"` keep their names and values, and existing
`CoordinateValidationError` handlers keep catching the failure.

## Deliberate exclusions (decisions, not oversights)

- **`Periodicity.from_string` keeps `ValueError`.** `_cli.py:128` passes it as
  an argparse `type=` callable; argparse turns `ValueError` into a usage error,
  so any other type would surface an unhandled traceback instead.
- **`Periodicity.unit()` keeps `ValueError`.** Unreachable internal invariant
  (an enum member without a unit), not user input validation.
- **CLI PET/KBDI periodicity raises stay `ValueError`.** "PET only supports
  monthly" is an inapplicable-argument combination, not an invalid
  `Periodicity` value, and #904 named only the `compute`/`indices` sites.
- **The exported-but-uncalled utils are kept, not deprecated.** The census in
  #904 is stale: `gregorian_length_as_366day` lost its `__spi__.py` caller when
  the legacy script was retired and `reshape_to_divs_years_months` was never
  listed, so six helpers now have no in-package caller rather than four. They
  keep working as documented public helpers, and `utils.__all__` now states
  that contract, because deprecating them would tell downstream users to stop
  calling helpers that have no in-library replacement. A follow-up can retire
  them once a removal release is chosen.
- **`BetaFeatureWarning` stays without a runtime emitter.** `emit_deprecation_warning`
  and `ClimateIndicesDeprecationWarning` keep their documented role for future
  deprecations. Warning on every beta xarray call would contradict #1012
  (RELEASE3-7, xarray API stability for 3.0.0), which is still open.

## Verification

- `uv run ruff check src/ tests/`, `uv run ruff format --check src/ tests/`, `uv run mypy src/`
- `uv run pytest` — 2176 passed, 423 deselected (benchmark/validation markers)
- `uv run pytest tests/test_review_scripts.py tests/test_release_integrity.py`
- `uv run sphinx-build -E -b html -W --keep-going docs docs/_build/html` (the CI docs gate; the new `{func}`/`{class}` refs resolve)

Tripwire updates that come with the type changes: the `HIERARCHY_CASES` and
`CATCHABILITY_CASES` tables in `tests/test_exceptions.py`, the
`valid_values` expectation in `tests/test_input_validation.py`, and the
`ValueError` expectations in `tests/test_compute.py`.

## Summary by Sourcery

Use the library’s specialized exceptions consistently for invalid periodicity values and missing named dimensions, with backward-compatible parent exception relationships.

Bug Fixes:
- Raise the library-defined periodicity and dimension-specific exceptions at the relevant validation sites instead of generic or parent exceptions.

Enhancements:
- Make PeriodicityError a specialization of InvalidArgumentError and DimensionMismatchError a specialization of CoordinateValidationError while preserving compatibility for handlers of the parent types.
- Document the changed exception behavior and retain uncalled utility helpers as supported public API.

Documentation:
- Document the 3.0.0 periodicity exception change, its migration guidance, and the refined missing-dimension exception hierarchy in troubleshooting and migration documentation.

Tests:
- Update exception hierarchy, validation, compute, and HDW tests to verify the specialized exception types and their context.

## Deferred

Review surfaced two pre-existing periodicity gaps this PR does not reach:
`transform_fitted_gamma`/`transform_fitted_pearson` return an all-missing input
before validation, and `_validate_array` only checks periodicity on its 1-D
branch, so `gamma_parameters`/`pearson_parameters` accept an invalid periodicity
for 2-D+ input. Both behave identically on `main`; fixing them adds error paths
where the library was previously silent, so they are tracked separately in #1039
rather than widened into this change.

### CODE REVIEW

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** [PR #1037 — refactor(exceptions): raise the periodicity and dimension errors the library defines (#904)](https://github.com/monocongo/climate_indices/pull/1037), branch `refactor/904-exported-surface` vs `origin/main`
- **Lanes run:** 5 read-only lanes dispatched in parallel via the `subagent` tool — public API/contracts, correctness/edge cases, tests, docs/repo gates, maintainability. All five returned successfully; every lane ran on the same pinned model (`deepseek/deepseek-flash`).
- **Validation (post-fix HEAD `df655fbe`):** `uv run ruff check src/ tests/` pass; `uv run ruff format --check src/ tests/` pass (95 files); `uv run mypy src/` pass (28 files); `uv run pytest` pass (2177 passed, 423 deselected); `uv run pytest tests/test_review_scripts.py tests/test_release_integrity.py` pass (45); `uv run sphinx-build -E -b html -W --keep-going docs docs/_build/html` pass.
- **Fix commits:** [f1e82de4 — refactor(exceptions): raise the periodicity and dimension errors the library defines (#904)](https://github.com/monocongo/climate_indices/commit/f1e82de48e14b9e6f4d371e1c566c6ba83b2882e), [df655fbe — fix(exceptions): align the wired periodicity errors with their docs and tests (#904)](https://github.com/monocongo/climate_indices/commit/df655fbe7ed44a5728fb081fd3777d569e2c1032)

### Findings and fix status

#### [P2] `docs/xarray_migration.md` still quoted the old class for the missing-time-dimension error — Fixed
- **Location:** `docs/xarray_migration.md` (line 445), mirrored into the committed `llms-full.txt` bundle (line 2888)
- **Origin:** API/contracts lane
- **Problem:** `validation.validate_time_dimension` now raises `DimensionMismatchError`, but the migration guide's Pitfall 1 snippet still read `CoordinateValidationError: Time dimension 'time' not found...`, contradicting the same snippet already corrected in `docs/troubleshooting.md`.
- **Applied fix:** part of `df655fbe` — the literal now reads `DimensionMismatchError`, and `llms-full.txt` was regenerated with `uv run scripts/generate_llms_txt.py` (`docs/xarray_migration.md` is in `FULL_FILES`, so the bundle gate in `tests/test_review_scripts.py` requires it).

#### [P2] `docs/troubleshooting.md` still quoted the old class for the periodicity error — Fixed
- **Location:** `docs/troubleshooting.md` (lines 704, 714)
- **Origin:** Docs lane
- **Problem:** The "Invalid periodicity" section showed `InvalidArgumentError: Invalid periodicity argument: weekly...` and a code comment saying `# raises InvalidArgumentError`, while `indices._validate_periodicity` now raises the `PeriodicityError` specialization. The type is still catchable as the parent, but the documented traceback no longer matched runtime.
- **Applied fix:** part of `df655fbe` — both lines name `PeriodicityError`; the remediation text is unchanged.

#### [P3] Migration entry named unexported helpers and omitted the public functions that changed — Fixed
- **Location:** `docs/deprecations/api-changes.md` (lines 101–107)
- **Origin:** Docs lane (also raised independently by the correctness and contracts lanes)
- **Problem:** The entry listed `reshape_values`, `gamma_parameters`, and `pearson_parameters`, which are absent from `compute.__all__`, so their `{func}` references render as plain text rather than links (the `-W` docs gate does not enable `nitpicky`), while `transform_fitted_gamma` and `transform_fitted_pearson` — exported, previously raising bare `ValueError`, and covered by the updated tests — went unmentioned.
- **Applied fix:** part of `df655fbe` — the entry now names `prepare_scaled`, `transform_fitted_gamma`, and `transform_fitted_pearson` with a resolving reference, keeps the unexported helpers as inline code, and states the `valid_values` attribute change (`"monthly, daily"` → `"Periodicity.monthly, Periodicity.daily"`) alongside the unchanged message text.

#### [P3] `periodicity_value` reported `None` on one path and `"None"` on five others — Fixed
- **Location:** `src/climate_indices/compute.py` (line 197, versus `:212`, `:326`, `:349`, `:1399`, `:1552`, `indices.py:126`)
- **Origin:** Correctness lane
- **Problem:** An omitted/`None` periodicity produced `err.periodicity_value is None` from `_validate_array`'s missing-periodicity branch and `== "None"` everywhere else, so handlers written against one encoding missed the other. The stringified form is the module's convention (`InvalidArgumentError.argument_value` is documented as stringified and pinned by `tests/test_input_validation.py`).
- **Applied fix:** part of `df655fbe` — the outlier now passes `periodicity_value=str(periodicity)`.

#### [P3] Exception hierarchy diagram omitted `PeriodicityError` under `InvalidArgumentError` — Fixed
- **Location:** `docs/troubleshooting.md` (lines 1345–1355)
- **Origin:** Contracts and docs lanes (both flagged it as pre-existing but in the diff)
- **Problem:** The diagram gained `DimensionMismatchError` under `CoordinateValidationError` but still showed `InvalidArgumentError` childless, even though the migration note now tells users to catch `PeriodicityError` (a child of it).
- **Applied fix:** part of `df655fbe` — `└── InvalidArgumentError` now carries `    └── PeriodicityError`.

#### [P3] HDW missing-`level_dim` test verified only the type and message — Fixed
- **Location:** `tests/test_fire.py` (lines 691–698)
- **Origin:** Tests lane
- **Problem:** The sibling site (`tests/test_validation.py`) asserts the `DimensionMismatchError` context attributes; the HDW site asserted only the class and message, so a regression dropping or mis-wiring `expected_dims`/`actual_dims` would stay green.
- **Applied fix:** part of `df655fbe` — the test now asserts `expected_dims == "level"`, `actual_dims == tuple(temperature.dims)`, `coordinate_name == "level"`, and `reason == "missing_dimension"`.

#### [P3] Invalid-periodicity rows did not pin which validator fired, and `gamma_parameters`' all-missing branch was untested — Fixed
- **Location:** `tests/test_compute.py` (lines 366–384; new tests at `:758–774`)
- **Origin:** Tests lane
- **Problem:** The `None` and `"unsupported_type"` rows asserted only `PeriodicityError`, so collapsing the two `_validate_array` branches would not fail them. Separately, the `ValueError` → `PeriodicityError` change in `gamma_parameters`' all-missing branch (`compute.py:1399`) had no test at all.
- **Applied fix:** part of `df655fbe` — the two rows are now context managers asserting their message and `periodicity_value`, a new test covers `gamma_parameters` with an all-NaN input and an invalid periodicity, and `test_reshape_values_rejects_unsupported_periodicity` covers the reshape guard. Note: converting those two rows also resolves two SonarCloud S9000 rows that #1031 lists for `tests/test_compute.py`; #1031's Step 0 re-validation will drop them.

#### [P3] Invalid periodicity still accepted on all-missing and 2-D+ `compute` inputs — Deferred to #1039
- **Location:** `src/climate_indices/compute.py` (`:823-827`, `:1687-1691`, `:207-212`)
- **Origin:** Correctness and docs lanes
- **Problem:** `transform_fitted_gamma`/`transform_fitted_pearson` return an all-missing input before `_validate_array` runs, and `_validate_array` only checks periodicity on its 1-D branch, so `gamma_parameters`/`pearson_parameters` ignore an invalid periodicity for 2-D+ `(years, periods)` input. Verified on this branch: both `transform_fitted_*` calls return `(24,)` for `periodicity="bogus"`, and `gamma_parameters` fits a 2-D input with `periodicity="monthly"`. Both gaps behave identically on `origin/main`, so this PR neither introduces nor widens them.
- **Applied fix:** **Deferred** — fixing them adds error paths where the library was previously silent, which is a behavior change outside a "wire the exceptions" PR. Filed as [#1039](https://github.com/monocongo/climate_indices/issues/1039) with the reproduction commands, the proposed shared guard, and acceptance criteria. The migration entry was reworded in `df655fbe` so it no longer implies full coverage of those functions.

### Verified non-issues
- `Periodicity.from_string` and `Periodicity.unit` keep `ValueError` deliberately: `_cli.py:128` passes `from_string` as an argparse `type=` callable, and argparse converts only `TypeError`/`ValueError` into a usage error, so the exclusion is load-bearing.
- `DimensionMismatchError`'s positional `(message, expected_dims, actual_dims)` signature is unchanged, its new `coordinate_name`/`reason` are keyword-only, pickle round-trips all context attributes (Dask workers rely on this), and no in-repo `except CoordinateValidationError` handler is broadened by the re-parenting.
- All migrated raises keep byte-identical message text and still emit their `reason=` structured log events before raising.
- `fit_and_standardize`'s gamma fallback (`compute.py:1918-1935`) no longer catches a `PeriodicityError` from the Pearson path, but every public index entry point validates periodicity before that call, so the narrowing is reachable only through direct `compute` use.

### Remaining blockers
None. No P0/P1 findings were raised; the single deferred item is a pre-existing gap tracked by #1039, so this PR is ready for review and merge by the maintainer.

File: src/climate_indices/compute.py [hunk 1]
```diff
@@ -1394,7 +1396,7 @@ def gamma_parameters(
         elif periodicity is Periodicity.daily:
             shape = (366,)
         else:
-            raise ValueError(f"Unsupported periodicity: {periodicity}")
+            raise PeriodicityError(f"Unsupported periodicity: {periodicity}", periodicity_value=str(periodicity))
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/compute.py:1399]: Addressed in df655fbe7ed44a5728fb081fd3777d569e2c1032: `tests/test_compute.py::test_gamma_parameters_all_missing_rejects_invalid_periodicity` calls `gamma_parameters` with an all-NaN array and `periodicity=None`, asserts `PeriodicityError` with `match="Unsupported periodicity"`, and pins `periodicity_value == "None"` — the concrete exception type and context the finding asked for. Passing locally.
